File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/holder/WebViewViewHolder.java
package com.glia.widgets.chat.adapter.holder;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.webkit.JavascriptInterface;
import android.webkit.WebSettings;
import android.webkit.WebView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.glia.androidsdk.chat.ChatMessage;
import com.glia.widgets.R;
import com.glia.widgets.chat.adapter.CustomCardMessage;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * The implementation of {@link CustomCardViewHolder} allows
 * displaying the card message as a WebView.
 * <p>
 * It can render HTML content if the {@link ChatMessage#getMetadata()}
 * has an HTML body with the <code>html</code> key.
 * <p>
 * <b>Metadata example:</b>
 * <pre>{@code
 * { "html": "
 *     <style>button {padding: 8px 16px;}p {color: green;}<\/style>
 *     <p>Lorem ipsum dolor sit amet<\/p>
 *     <p>
 *         <button
 *              type=\"button\"
 *              onClick=\"sendResponse('Cancel', 'response_cancel')\">
 *              Cancel
 *         <\/button>
 *         <button
 *              type=\"button\"
 *              onClick=\"sendResponse('OK', 'response_ok')\">
 *              OK
 *         <\/button>
 *     <\/p>
 * " }
 * }<pre/>
 * @see CustomCardViewHolder
 */
public class WebViewViewHolder extends CustomCardViewHolder {
    private static final String MIME_TYPE = "text/html";
    private static final String ENCODING = "UTF-8";
    private static final String METADATA_KEY = "html";
    private static final String JS_SCRIPT =
            "<script type=\"text/javascript\">" +
                    "function sendResponse(text, value){Glia.response(text, value);}" +
                    "function callMobileAction(action){Glia.action(action);}" +
            "</script>";

    private final WebView webView;

    @Nullable
    private ResponseCallback responseCallback;

    @Nullable
    private MobileActionCallback mobileActionCallback;

    @SuppressLint("SetJavaScriptEnabled")
    public WebViewViewHolder(@NonNull ViewGroup parent) {
        super(LayoutInflater.from(parent.getContext()).inflate(R.layout.web_view_layout, parent, false));

        webView = itemView.findViewById(R.id.web_view);
        WebSettings webViewSettings = webView.getSettings();
        webViewSettings.setJavaScriptEnabled(true);
        webViewSettings.setAllowFileAccess(false);
        webView.addJavascriptInterface(new JavaScriptInterface(), "Glia");
    }

    /**
     * Register a callback to be invoked when
     * the JS method {@code callMobileAction(String)} called inside WebView message.
     * @param mobileActionCallback the callback that will run
     */
    public void setMobileActionCallback(@Nullable MobileActionCallback mobileActionCallback) {
        this.mobileActionCallback = mobileActionCallback;
    }

    /**
     * @see CustomCardViewHolder#bind(CustomCardMessage, ResponseCallback)
     */
    @Override
    public void bind(@NonNull CustomCardMessage message, @NonNull ResponseCallback callback) {
        responseCallback = callback;

        JSONObject metadata = message.getMetadata();
        if (metadata != null) {
            try {
                String html = metadata.getString(METADATA_KEY);
                webView.loadDataWithBaseURL("", html + JS_SCRIPT, MIME_TYPE, ENCODING, "");
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * Allows checking if the message can be displayed using {@link WebViewViewHolder}.
     * @param message the chat message with metadata.
     * @return true if the message metadata has the <code>html</code> key.
     */
    public static boolean isWebViewType(@NonNull ChatMessage message) {
        JSONObject metadata = message.getMetadata();
        if (metadata == null || metadata.length() == 0) {
            return false;
        }
        return metadata.has(METADATA_KEY);
    }

    /**
     * Allows checking if the message can be displayed using {@link WebViewViewHolder}.
     * @param message the chat message with metadata.
     * @return true if the message metadata has the <code>html</code> key.
     */
    public static boolean isWebViewType(@NonNull CustomCardMessage message) {
        JSONObject metadata = message.getMetadata();
        if (metadata == null || metadata.length() == 0) {
            return false;
        }
        return metadata.has(METADATA_KEY);
    }

    private class JavaScriptInterface {

        @JavascriptInterface
        public void response(String text, String value) {
            if (responseCallback != null) {
                responseCallback.sendResponse(text, value);
            }
        }

        @JavascriptInterface
        public void action(String action) {
            if (mobileActionCallback != null) {
                mobileActionCallback.onMobileAction(action);
            }
        }
    }

    /**
     * Allows sending the action from a custom card.
     */
    public interface MobileActionCallback {
        /**
         * @param action the value of an action.
         */
        void onMobileAction(String action);
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/holder/WebViewViewHolder.kt
package com.glia.widgets.chat.adapter.holder

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import android.webkit.JavascriptInterface
import android.webkit.WebView
import com.glia.androidsdk.chat.ChatMessage
import com.glia.widgets.R
import com.glia.widgets.chat.adapter.CustomCardMessage
import org.json.JSONException

/**
 * The implementation of [CustomCardViewHolder] allows displaying the card message as a WebView.
 * <p>
 * It can render HTML content if the [ChatMessage.metadata] has an HTML body with the `html` key.
 * <p>
 * <b>Metadata example:</b>
 * <pre>{@code
 * { "html": "
 *     <style>button {padding: 8px 16px;}p {color: green;}<\/style>
 *     <p>Lorem ipsum dolor sit amet<\/p>
 *     <p>
 *         <button
 *              type=\"button\"
 *              onClick=\"sendResponse('Cancel', 'response_cancel')\">
 *              Cancel
 *         <\/button>
 *         <button
 *              type=\"button\"
 *              onClick=\"sendResponse('OK', 'response_ok')\">
 *              OK
 *         <\/button>
 *     <\/p>
 * " }
 * }<pre/>
 * @see CustomCardViewHolder
</pre> */
class WebViewViewHolder @SuppressLint("SetJavaScriptEnabled") constructor(parent: ViewGroup) :
    CustomCardViewHolder(
        LayoutInflater.from(parent.context).inflate(R.layout.web_view_layout, parent, false)
    ) {
    private val webView: WebView = itemView.findViewById(R.id.web_view)
    private var responseCallback: ResponseCallback? = null
    private var mobileActionCallback: MobileActionCallback? = null

    init {
        val webViewSettings = webView.settings
        webViewSettings.javaScriptEnabled = true
        webViewSettings.allowFileAccess = false
        webView.addJavascriptInterface(JavaScriptInterface(), "Glia")
    }

    /**
     * Register a callback to be invoked when
     * the JS method `callMobileAction(String)` called inside WebView message.
     * @param mobileActionCallback the callback that will run
     */
    fun setMobileActionCallback(mobileActionCallback: MobileActionCallback?) {
        this.mobileActionCallback = mobileActionCallback
    }

    /**
     * @see CustomCardViewHolder.bind
     */
    override fun bind(message: CustomCardMessage, callback: ResponseCallback) {
        responseCallback = callback

        val metadata = message.metadata
        if (metadata != null) {
            try {
                val html = metadata.getString(METADATA_KEY)
                webView.loadDataWithBaseURL("", html + JS_SCRIPT, MIME_TYPE, ENCODING, "")
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
    }

    private inner class JavaScriptInterface {
        @JavascriptInterface
        fun response(text: String, value: String) {
            responseCallback?.sendResponse(text, value)
        }

        @JavascriptInterface
        fun action(action: String) {
            mobileActionCallback?.onMobileAction(action)
        }
    }

    /**
     * Allows sending the action from a custom card.
     */
    fun interface MobileActionCallback {
        /**
         * @param action the value of an action.
         */
        fun onMobileAction(action: String)
    }

    /**
     * Companion object containing constants and utility methods for WebViewViewHolder.
     * Provides metadata keys, MIME type, encoding, and helper functions to determine
     * if a message can be displayed using a WebView.
     */
    companion object {
        private const val MIME_TYPE = "text/html"
        private const val ENCODING = "UTF-8"
        private const val METADATA_KEY = "html"
        private const val JS_SCRIPT = "<script type=\"text/javascript\">" +
            "function sendResponse(text, value){Glia.response(text, value);}" +
            "function callMobileAction(action){Glia.action(action);}" +
            "</script>"

        /**
         * Allows checking if the message can be displayed using [WebViewViewHolder].
         * @param message the chat message with metadata.
         * @return true if the message metadata has the `html` key.
         */
        fun isWebViewType(message: ChatMessage): Boolean {
            val metadata = message.metadata
            if (metadata == null || metadata.length() == 0) {
                return false
            }
            return metadata.has(METADATA_KEY)
        }

        /**
         * Allows checking if the message can be displayed using [WebViewViewHolder].
         * @param message the chat message with metadata.
         * @return true if the message metadata has the `html` key.
         */
        fun isWebViewType(message: CustomCardMessage): Boolean {
            val metadata = message.metadata
            if (metadata == null || metadata.length() == 0) {
                return false
            }
            return metadata.has(METADATA_KEY)
        }
    }
}
